File: 2024/day11/solution.py
#!/usr/bin/env python3
import time
from collections import Counter

DEBUG = True
EXAMPLE = False


def run(items: list[int], iterations: int) -> int:
  counts = Counter(items)
  for iteration in range(iterations):
    if DEBUG: i_started = time.monotonic()
    new_counts: Counter[int] = Counter()
    for value, count in counts.items():
      if value == 0:
        new_counts[1] += count
      else:
        value_str = str(value)
        value_str_len = len(value_str)
        if value_str_len % 2 == 0:
          new_counts[int(value_str[:value_str_len // 2])] += count
          new_counts[int(value_str[value_str_len // 2:])] += count
        else:
          new_counts[value * 2024] += count
    if DEBUG:
      i_elapsed = time.monotonic() - i_started
      count_total = sum(new_counts.values())
      print(f'{iteration}: {i_elapsed:.3f}s {len(counts.keys())} -> {len(new_counts.keys())} {count_total=}')
    counts = new_counts
  return sum(counts.values())


with open('example.txt' if EXAMPLE else 'input.txt', 'r') as f:
  text = f.read()
if DEBUG: print(f'{text=}')
items = list(map(int, text.split()))
if DEBUG: print(f'{items=}')
print(f'part 1 {run(items, 25)}')
print(f'part 2 {run(items, 75)}')

File: 2024/day11/solution.ts
#!/usr/bin/env -S deno --allow-read

const DEBUG = false;
const EXAMPLE = false;
const ITERATIONS = 75;

const text = await Deno.readTextFile(EXAMPLE ? 'example.txt' : 'input.txt');
if (DEBUG) console.debug({ text });

class NumberCounter {
  private counts: Record<string, number> = {};
  private total: number = 0;
  constructor(values: number[] | string[] = []) {
    for (const value of values) {
      this.add(value);
    }
  }
  add(value: number | string, count: number = 1): void {
    const valueString = typeof value === 'string' ? value : String(value);
    if (DEBUG) console.debug({ value, valueString, count });
    this.counts[valueString] = (this.counts[valueString] ?? 0) + count;
    this.total += count;
  }
  totalCount(): number {
    return this.total;
  }
  entries(): [string, number][] {
    return Object.entries(this.counts);
  }
}

let counts = new NumberCounter(text.split(/\s+/));
if (DEBUG) console.debug({ counts });

for (let iteration = 0; iteration < ITERATIONS; iteration++) {
  const newCounts = new NumberCounter();
  for (const [valueStr, count] of counts.entries()) {
    if (valueStr === '0') {
      newCounts.add('1', count);
    } else if (valueStr.length % 2 == 0) {
      newCounts.add(valueStr.slice(0, valueStr.length / 2), count);
      newCounts.add(parseInt(valueStr.slice(valueStr.length / 2)), count);
    } else {
      newCounts.add(parseInt(valueStr) * 2024, count);
    }
  }
  const totalCount = newCounts.totalCount();
  if (DEBUG) {
    console.debug({ iteration, counts, newCounts, totalCount });
  } else {
    console.log({ iteration, totalCount });
  }
  counts = newCounts;
}
